orttraining/orttraining/training_ops/cuda/nn/dropout.cc

```diff
@@ -102,5 +102,112 @@ Status DropoutGrad<T1, T2>::ComputeInternal(OpKernelContext* context) const {
+
+ //Get bias_data
+  const Tensor* bias = context->Input<Tensor>(1);
+  if (bias == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "Bias input of BiasDropout is not available.");
+  const TensorShape& bias_shape = bias->Shape();
+  if (bias_shape.NumDimensions() != 1) {
+    return Status(common::ONNXRUNTIME, common::FAIL, "Bias input is not a 1D tensor.");
+  }
+  const int64_t dim = bias_shape[0];
+  if (dim != x_shape.GetDims().back()){
+    return Status(common::ONNXRUNTIME, common::FAIL, "Bias' dimension doesn't match input's last dimension.");
+  }
+  auto bias_data = reinterpret_cast<const CudaT*>(bias->template Data<T1>());
+
+  //Get residual_data
+  const Tensor* residual = context->Input<Tensor>(2);
+  const CudaT* residual_data = nullptr;
+  if (residual != nullptr) {
+    const TensorShape& residual_shape = residual->Shape();
+    if (residual_shape != x_shape) {
+      return Status(common::ONNXRUNTIME, common::FAIL, "Residual input shape does not match X input shape.");
+    }
+    residual_data = reinterpret_cast<const CudaT*>(residual->template Data<T1>());
+  }
+
+  //Get Y_data
+  auto Y = context->Output(0, x_shape);
+  auto Y_data = reinterpret_cast<CudaT*>(Y->template MutableData<T1>());
+
+  //Get mask_data
+  auto mask = context->Output(1, x_shape);
+  ORT_ENFORCE(!mask || mask->Shape().Size() == N);
+
+  //Get the ratio_data
+  float ratio_data;
+  auto ratio = context->Input<Tensor>(3);
+
+  static_assert(std::is_same<T2, MLFloat16>::value || std::is_same<T2, float>::value || std::is_same<T2, double>::value,
+                "T2 must be float16 or float or double");
+
+  if (ratio) {
+    ratio_data = static_cast<float>(*(ratio->template Data<T2>()));
+  } else {
+    ratio_data = default_ratio_;
+  }
+  ORT_ENFORCE(ratio_data >= 0.0f && ratio_data < 1.0f);
+
+  // const Tensor* training_mode = context->Input<Tensor>(4);
+  // //Check for inference mode.
+  // if (training_mode == nullptr || *(training_mode->Data<bool>()) == false) {
+  //   if (Y_data != X_data) {
+  //     CUDA_CALL_THROW(cudaMemcpyAsync(Y_data, X_data, N * sizeof(T1), cudaMemcpyDeviceToDevice));
+  //   }
+
+  //   // If mask is requested, return all 1s.
+  //   if (mask != nullptr) {
+  //     ORT_ENFORCE(cudaMemset(mask->MutableData<bool>(), true, N * sizeof(bool)) == cudaSuccess);
+  //   }
+
+  //   return Status::OK();
+  // }
+
+  IAllocatorUniquePtr<bool> temp_mask_buffer{};  // buffer to use if mask is not provided
+  bool* const mask_data = [this, N, mask, &temp_mask_buffer]() {
+    if (mask) return mask->MutableData<bool>();
+    temp_mask_buffer = GetScratchBuffer<bool>(N);
+    return temp_mask_buffer.get();
+  }();
+
+  const fast_divmod fdm_dim(gsl::narrow_cast<int>(dim));
+  PhiloxGenerator& generator = generator_ != nullptr ? *generator_.get() : PhiloxGenerator::Default();
+  BiasDropoutKernelImpl(GetDeviceProp(), N, fdm_dim, ratio_data, generator, X_data, bias_data, residual_data, Y_data, mask_data);
```

Comment:
>BiasDropoutKernelImpl [](start = 2, length = 21)

can dispatch to templated kernel launch function here with MLTypeCallDispatcher

Reply:
didn't see the change here - with the dispatch we should be able to make the BiasDropout class not templated. the templated tensor data access can happen inside the function that is dispatched to.